miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultInstrumentService.java

```diff
@@ -71,9 +77,74 @@ public long create(Instrument instrument) throws IOException {
     return save(instrument);
   }
 
+  @Override
+  public void addServiceRecord(ServiceRecord record, Instrument instrument) throws IOException {
+    // originally located at HibernateServiceRecordDao
+    if (instrument.getDateCommissioned() != null) {
+      throw new IOException("Cannot add service records to a retired instrument!");
+    }
+    long recordId = serviceRecordService.create(record);
+    ServiceRecord serviceRecord = serviceRecordService.get(recordId);
+
+    // add saved service record
+    instrument.getServiceRecords().add(serviceRecord);
+
+    // save instrument
+    save(instrument);
+  }
+
+  @Override
+  public void removeServiceRecord(ServiceRecord record, Instrument instrument) throws IOException {
+    if (!instrument.getServiceRecords().contains(record)) {
```

Comment:
You'll need to first pull the instrument from the database (to ensure you're saving the managed object). What do you think about the following?
```
Instrument managedInstrument = get(instrument.getId());
ServiceRecord managedServiceRecord = managedInstrument.getServiceRecords()
    .stream()
    .filter(sr -> sr.getId() == record.getId())
    .findFirst()
    .orElseNull();
if (managedServiceRecord == null) {
   throw new ValidationException(
      String.format("Service record '%s' is not associated with instrument %s; does the service record exist?", 
        managedServiceRecord.getId(), 
        managedInstrument.getName()
    ));
}
managedInstrument.getServiceRecords().remove(managedServiceRecord);
save(managedInstrument);

serviceRecordService.delete(managedServiceRecord);
```